src/qt/listctrl.cpp

```diff
@@ -141,96 +144,868 @@ class wxQtTreeWidget : public wxQtEventSignalHandler< QTreeWidget, wxListCtrl >
+            ? wxIMAGE_LIST_SMALL
+            : wxIMAGE_LIST_NORMAL;
+        return m_listCtrl->GetImageList(requiredList);
+    }
+
+    wxListCtrl *GetListCtrl() const
+    {
+        return m_listCtrl;
+    }
+
+    bool IsSelected(const QModelIndex& index) const
+    {
+        QModelIndexList indices = m_view->selectionModel()->selectedIndexes();
+        return indices.contains(index);
+    }
+
+    void CopySelectStatusToItem(wxListItem& info, const int row, const int col)
+    {
+        if (info.m_mask & wxLIST_MASK_STATE)
+        {
+            info.m_state = wxLIST_STATE_DONTCARE;
+            if (info.m_stateMask & wxLIST_STATE_FOCUSED)
+            {
+                if (m_view->currentIndex().row() == row)
+                    info.m_state |= wxLIST_STATE_FOCUSED;
+            }
+            if (info.m_stateMask & wxLIST_STATE_SELECTED)
+            {
+                if (IsSelected(index(row, col)))
+                    info.m_state |= wxLIST_STATE_SELECTED;
+            }
+        }
+    }
+
+private:
+    template <typename  T>
+    static void eraseFromContainer(T &container, int start_index, int count)
+    {
+
+        typename T::iterator first = container.begin();
+        std::advance(first, start_index);
+
+        typename T::iterator last = first;
+        std::advance(last, count);
+
+        container.erase(first, last);
+    }
+
+    struct ColumnItem
+    {
+        ColumnItem() :
+            m_align(Qt::AlignLeft),
+            m_image(-1),
+            m_selectedImage(-1)
+        {
+        }
+
+        QString m_label;
+        QColor m_backgroundColour;
+        QColor m_textColour;
+        QFont m_font;
+        Qt::AlignmentFlag m_align;
+        int m_image;
+        int m_selectedImage;
+    };
+
+    struct RowItem
+    {
+        RowItem() :
+            m_data(NULL),
+            m_checked(false)
+        {
+        }
+
+        RowItem(int columnCount ) :
+            m_columns(columnCount),
+            m_data(NULL),
+            m_checked(false)
+        {
+        }
+
+        const ColumnItem &operator[](int index) const
+        {
+            return m_columns[index];
+        }
+
+        ColumnItem &operator[](int index)
+        {
+            return m_columns[index];
+        }
+
+        std::vector<ColumnItem> m_columns;
+        void *m_data;
+        bool m_checked;
+
+    };
+
+    struct CompareAdapter
+    {
+        CompareAdapter(wxListCtrlCompare fn, wxIntPtr data) :
+            m_fn(fn),
+            m_data(data)
+        {
+        }
+
+        bool operator()(const RowItem &lhs, const RowItem &rhs) const
+        {
+            return m_fn(
+                reinterpret_cast<wxIntPtr>(lhs.m_data),
+                reinterpret_cast<wxIntPtr>(rhs.m_data),
+                m_data) < 0;
+        }
+
+        wxListCtrlCompare m_fn;
+        wxIntPtr m_data;
+    };
+
+    std::vector<ColumnItem> m_headers;
+    std::vector<RowItem> m_rows;
```

Comment:
It doesn't look right to have this in the base class, when the virtual list model doesn't need it. Could this be refactored into `wxQtListModelBase` and `wxQt{Concrete,Virtual}ListModel`s deriving from it?

Reply:
Before I make this change,  could you clarify some of the behaviour.    Is it valid to enable check boxes for virtual list controls?

If so,  is it the responsibility of wxListCtrl or the derived type to maintain the state of which items are checked?

Reply:
Unfortunately, even though in principle checkboxes could well work for virtual controls, the native MSW control doesn't support combining `LVS_EX_CHECKBOXES` with `LVS_OWNERDATA`, so this can't work there and so it doesn't make much sense to support this in the other ports neither (even though it's not really forbidden, of course, it's just that few people will use it if it doesn't work in wxMSW and, also, this would require some extra work, of course).